### PR TITLE
impl: libp2p metrics

### PR DESCRIPTION
**Pull request by RolandSherwin**
Depends on #706 
- Enables libp2p metrics. Example of the metrics collected https://kademlia-exporter.max-inden.de/d/Pfr0Fj6Mk/rust-libp2p?orgId=1&refresh=30s
- It is enabled by default for the node.
- Can be extended to support more custom metrics.

### Summary generated by Reviewpad on 04 Sep 23 14:47 UTC
This pull request includes changes to multiple files. Here is an overview of the changes:

1. `SwarmDriver` in `sn_networking` crate:
The implementation of the `SwarmDriver` struct and its associated methods have been modified. Changes include:
- Importing and using the `driver::SwarmDriver` trait.
- Importing and using the `libp2p_metrics::Recorder` trait.
- Removing the `NodeBehaviour` struct.
- Updating the `NodeEvent` enum to include additional events.
- Updating the `SwarmDriver` struct to include a `network_metrics` field of type `Recorder`.
- Updating the `handle_msg` method to handle request/response messages using the `request_response` crate.
- Updating the `handle_kad_event` method to handle Kademlia events.

2. Addition of `metrics_service.rs`:
A new file called `metrics_service.rs` has been added. This file implements a metrics service for the SAFE Network Software. It defines a `MetricService` struct and a `MakeMetricService` struct. The `MetricService` struct implements the `Service` trait from the `hyper` crate and provides functions for handling HTTP requests. The `MakeMetricService` struct is responsible for creating instances of the `MetricService` struct. The file also contains a `metrics_server` function that runs the metrics service on `localhost`.

3. Changes to `Cargo.toml`:
The `Cargo.toml` file has been modified to include new dependencies and update existing ones. Changes include:
- Adding the dependency `libp2p-metrics` with version `0.13.1` and features `identify` and `kad`.
- Adding the dependency `prometheus-client` with version `0.21.2`.
- Updating the dependency `hyper` to version `0.14` and adding features `server`, `tcp`, and `http1`.

4. Changes to `cmd.rs`:
The `cmd.rs` file has been modified to include changes related to imports and function removal. Changes include:
- Adding the import statement `use crate::driver::SwarmDriver`.
- Removing the import statement `use libp2p::swarm::dial_opts::DialOpts`.
- Removing the functions `dial` and `dial_with_opts`.
- Removing some code related to dialing with a peer ID.

5. Other changes:
- In a certain file, changes include removing the import of the `rand::Rng` module, removing the definitions of specific constants, adjusting the implementation of the `Default` trait for a struct, removing unused imports, and removing some comments.
- Additionally, the file `sn_networking/src/msg/mod.rs` has been deleted.

Please review these changes to ensure they are correct and meet the requirements of the codebase. Let me know if you need any further clarification!

### CODE REVIEW

File: sn_networking/src/event.rs [hunk 1]
```diff
@@ -284,6 +292,28 @@ impl SwarmDriver {
...
+                autonat::Event::StatusChanged { old, new } => {
+                    info!("AutoNAT status changed: {old:?} -> {new:?}");
+                    self.send_event(NetworkEvent::NatStatusChanged(new.clone()));
+
+                    match new {
+                        NatStatus::Public(_addr) => {
+                            // In theory, we could actively push our address to our peers now. But, which peers? All of them?
+                            // Or, should we just wait and let Identify do it on its own? But, what if we are not connected
+                            // to any peers anymore? (E.g., our connections timed out etc)
+                            // let all_peers: Vec<_> = self.swarm.connected_peers().cloned().collect();
```
Comment (joshuef) [inline sn_networking/src/event.rs:307]: should this be uncommented? or removed?

**Comment by joshuef** [comment]
How do we access these metrics? Do we need tostart a custom server for this? does each node run a server to access metrics?

**Review by joshuef — CHANGES_REQUESTED** [review]
I think `metrics` here confuses with out logging metrics output. I'd say we want a clearer name for this feature.

`libp2p-metrics`, `network-metrics` ? `something-else` ?

**Comment by joshuef** [comment]
I'm also torn on this being on by default atm...

**Comment by RolandSherwin** [comment]
> How do we access these metrics? Do we need tostart a custom server for this? does each node run a server to access metrics?

Yes, each node runs a separate hyper server on a random port. This returns the `OpenMetrics` response type that can be fed into grafana / other tools.

Also, I will push a change to update the feature name and remove it from the default features list.

**Review by joshuef — APPROVED** [review]

**Pull request merged**